### ISSUE TITLE
compilation with sancov and multiple codegen units results in SIGSEGV: invalid memory reference; nightly-2017-12-26 regression

### ISSUE THREAD

**Issue by frewsxcv** [rust-lang/rust#47071]
```
git clone https://github.com/rust-fuzz/targets.git
cd targets
export RUSTFLAGS="-C codegen-units=16 -C passes=sancov -C llvm-args=-sanitizer-coverage-level=3 -Z sanitizer=address -C opt-level=3"
cargo build --verbose
```

```
error: Could not compile `getopts`.

Caused by:
  process didn't exit successfully: `rustc --crate-name getopts /Users/corey/.cargo/registry/src/github.com-1ecc6299db9ec823/getopts-0.2.15/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=e4fe6dbb6aba3ffd -C extra-filename=-e4fe6dbb6aba3ffd --out-dir /private/tmp/targets/target/debug/deps -L dependency=/private/tmp/targets/target/debug/deps --cap-lints allow -C passes=sancov -C llvm-args=-sanitizer-coverage-level=3 -Z sanitizer=address -C opt-level=3` (signal: 11, SIGSEGV: invalid memory reference)
```

EDIT: updated `RUSTFLAGS` to include `-C codegen-units=16` since rust-fuzz/targets `master` was updated to use `-C codegen-units=1`

**Comment by frewsxcv** [comment]
if i add `-C codegen-units=1` to `RUSTFLAGS`, it compiles fine. so i'm going to assume this started happening after https://github.com/rust-lang/rust/pull/46910 merged

**Comment by frewsxcv** [comment]
another possibly related issue https://github.com/rust-lang/rust/issues/45220

**Comment by frewsxcv** [comment]
this might be the issue upstream https://bugs.llvm.org/show_bug.cgi?id=32700

**Comment by nikomatsakis** [comment]
So, am I correctly inferring from this that this is (most likely) not a rustc bug, but rather a bug in LLVM?

**Comment by nikomatsakis** [comment]
@frewsxcv can you get a backtrace, by any chance?

**Comment by nikomatsakis** [comment]
Even better, can you test with a local build, that has LLVM asseritons enabled?

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by frewsxcv** [comment]
i tried setting `RUST_BACKTRACE=1` but no backtrace gets printed upon compilation failure. similarly, i compiled rustc by with `assertions = true` in rustc's config.toml and couldn't seem to get llvm assertions enabled either.

**Comment by pnkfelix** [comment]
triage: Leaving at P-medium (for reasons noted by @nikomatsakis above). Might be good for a second set of eyes beyond @frewsxcv to double-check about running on a local build with llvm assertions enabled.

**Comment by Mark-Simulacrum** [comment]
This isn't an error with getopts, but I think this may help.

I also set `LSAN_OPTIONS=verbosity=1:log_threads=1`. Otherwise followed directions in original post.

```
error: failed to run custom build command for `pulldown-cmark v0.0.15 (https://github.com/google/pulldown-cmark.git#13918e7d)`
process didn't exit successfully: `/home/tester/targets/target/debug/build/pulldown-cmark-04a92855a8df93af/build-script-build` (exit code: 1)
--- stderr
==21919==AddressSanitizer: failed to intercept '__isoc99_printf'
==21919==AddressSanitizer: failed to intercept '__isoc99_sprintf'
==21919==AddressSanitizer: failed to intercept '__isoc99_snprintf'
==21919==AddressSanitizer: failed to intercept '__isoc99_fprintf'
==21919==AddressSanitizer: failed to intercept '__isoc99_vprintf'
==21919==AddressSanitizer: failed to intercept '__isoc99_vsprintf'
==21919==AddressSanitizer: failed to intercept '__isoc99_vsnprintf'
==21919==AddressSanitizer: failed to intercept '__isoc99_vfprintf'
==21919==AddressSanitizer: failed to intercept '__cxa_throw'
==21919==AddressSanitizer: libc interceptors initialized
|| `[0x10007fff8000, 0x7fffffffffff]` || HighMem    ||
|| `[0x02008fff7000, 0x10007fff7fff]` || HighShadow ||
|| `[0x00008fff7000, 0x02008fff6fff]` || ShadowGap  ||
|| `[0x00007fff8000, 0x00008fff6fff]` || LowShadow  ||
|| `[0x000000000000, 0x00007fff7fff]` || LowMem     ||
MemToShadow(shadow): 0x00008fff7000 0x000091ff6dff 0x004091ff6e00 0x02008fff6fff
redzone=16
max_redzone=2048
quarantine_size_mb=256M
thread_local_quarantine_size_kb=1024K
malloc_context_size=30
SHADOW_SCALE: 3
SHADOW_GRANULARITY: 8
SHADOW_OFFSET: 0x7fff8000
==21919==Installed the sigaction for signal 11
==21919==Installed the sigaction for signal 7
==21919==Installed the sigaction for signal 8
==21919==T0: stack [0x7ffe66357000,0x7ffe66b57000) size 0x800000; local=0x7ffe66b54e70
==21919==AddressSanitizer Init done
==21938==Could not attach to thread 21919 (errno 1).
==21938==Failed suspending threads.
==21919==LeakSanitizer has encountered a fatal error.
==21919==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==21919==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
```

**Comment by alexcrichton** [comment]
Removing from the 1.25 milestone as this is P-medium and also looks like it uses unstable features

**Comment by steveklabnik** [comment]
Triage: I tried to reproduce but

```
error: AddressSanitizer only works with the `x86_64-unknown-linux-gnu` or `x86_64-apple-darwin` or `x86_64-fuchsia` or `aarch64-fuchsia` target
```

**Comment by nikomatsakis** [comment]
I wonder if @tmiasko has any insight here, they've been effectively maintaining sanitizers at this point.

**Comment by mati865** [comment]
Error on Linux:
```
  = note: /usr/bin/ld: __sancov_guards has both ordered [`__sancov_guards' in /tmp/targets/target/debug/deps/libsyn-70b79fe435ab6acd.rlib(syn-70b79fe435ab6acd.syn.6z3kueqo-cgu.14.rcgu.o)] and unordered [`__sancov_guards' in /tmp/targets/target/debug/deps/libsyn-70b79fe435ab6acd.rlib(syn-70b79fe435ab6acd.syn.6z3kueqo-cgu.14.rcgu.o)] sections
          /usr/bin/ld: final link failed: Bad value
          collect2: error: ld returned 1 exit status
```

**Comment by tmiasko** [comment]
@mati865 the linking issue would be #53945 (SanitizerCoverage has poor compatibility with some linkers & dead code stripping & LTO).

@frewsxcv is this still an issue for you?

It builds correctly for me with gold linker:

```shell
export RUSTFLAGS="-Clink-arg=-fuse-ld=gold -C codegen-units=16 -C passes=sancov -C llvm-args=-sanitizer-coverage-level=3 -Z sanitizer=address -C opt-level=3"
cargo build --target x86_64-unknown-linux-gnu
```

Possibly a duplicate of already closed #47953 (also about SanitizerCoverage and opened around similar time)?
